### PR TITLE
Refactored DFS support

### PR DESCRIPTION
**Pull request by hierynomus**

### CODE REVIEW

**Comment by hierynomus** [comment]
Based on #33

**Comment by hierynomus** [comment]
@ckherrmann @calibanatspace @gidravlic @sbellan Could you verify this PR to see whether it works  "as expected"?

**Comment by gidravlic** [comment]
Hi hierynomus! You want that i make tests with ckherrmann-dfs branch? If only this, i'll test this tomorrow. My working time ends :)

**Comment by hierynomus** [comment]
@gidravlic Yes, if you could do that, I would greatly appreciate it. Thanks a lot!
Because it is such a big change, I would welcome a few extra pairs of eyes (and tests) to see whether the DFS support is in working order now.

**Comment by gidravlic** [comment]
OK! I'll do it until 12:00 Helsinki, Riga time.

**Comment by gidravlic** [comment]
1) After switch to ckherrmann-dfs Netbeans 8.2 IDE told suggest that SMB2SessionSetup the third parameter in constructor EnumSet<ConnectionInfo.GlobalCapability> capabilities (cannot find symbol) 
image
2) Respectively in Connection class in method authenticationRound
`SMB2SessionSetup req = new SMB2SessionSetup(connectionInfo.getNegotiatedProtocol().getDialect(), EnumSet.of(SMB2_NEGOTIATE_SIGNING_ENABLED), 
                connectionInfo.getClientCapabilities());`
image

**Comment by hierynomus** [comment]
If I look at the code on github on the branch, it should be `EnumSet<SMB2GlobalCapability> capabilities`. I think the switch to the branch is incorrect. Did you cleanly clone from this repo?

**Comment by gidravlic** [comment]
No, i just switch branch. Ok, i'll cleanly clone repo.

**Comment by hierynomus** [comment]
Please do, thanks, it seems there is some leftover code somewhere in your tree. Also travis claims this branch builds fine ;)

**Comment by gidravlic** [comment]
Yes, you're right. Clean clone helped, nо errors and warnings :)

**Comment by hierynomus** [comment]
Good to hear @gidravlic, let me know once you've tested it :)

**Comment by gidravlic** [comment]
Hi hierynomus! I'm some trouble to test new smbj implementation with DFS functionality, because i think that I do not understand how correctly use new SMBClient implimentation. I thought it would be like this:
```java
SMBClient client = new SMBClient("\\somedfs\share\directory1\directory2");
client.authenticate(new AuthenticationContext(user, password, domain)
```
but it is not :)

In previous version smbj i create wrapper for SMPClient:
```java
public class SMBClientEx extends SMBClient {

    private final static Logger log = getLogger(SMBClientEx.class);

    private final SmbUrl smbUrl;
    private final Connection connection;
    private Session session;
    private DiskShare diskShare;
    private PrinterShare printerShare;

    /**
     *
     * @param url
     * @throws MalformedURLException
     * @throws IOException
     */
    public SMBClientEx(String url) throws MalformedURLException, IOException {
        super();
        smbUrl = new SmbUrl(url);
        connection = connect(smbUrl.getHost());
    }

    /**
     *
     * @return connection url
     */
    public SmbUrl getSmbUrl() {
        return smbUrl;
    }

    /**
     *
     * @return share name
     */
    public String getShare() {
        return smbUrl.getShare();
    }

    /**
     *
     * @return share path
     */
    public String getPath() {
        return smbUrl.getPath();
    }

    /**
     *
     * @return port
     */
    public int getPort() {
        return smbUrl.getPort();
    }

    /**
     *
     * @return protocol
     */
    public String getProtocol() {
        return smbUrl.getProtocol();
    }

    /**
     *
     * @return host name
     */
    public String getHost() {
        return smbUrl.getHost();
    }

    /**
     *
     * @return disk share
     */
    public DiskShare getDiskShare() {
        return diskShare;
    }

    /**
     *
     * @return printer share
     */
    public PrinterShare getPrinterShare() {
        return printerShare;
    }

    /**
     *
     * @param authenticationContext
     */
    public void authenticate(AuthenticationContext authenticationContext) {
        session = connection.authenticate(authenticationContext);
        connectShare();
    }

    /**
     *
     */
    private void connectShare() {
        Share share = session.connectShare(smbUrl.getShare());
        if (share instanceof DiskShare) {
            diskShare = (DiskShare) share;
        } else if (share instanceof PrinterShare) {
            printerShare = (PrinterShare) share;
        }
    }

    /**
     *
     * @param fileInfo
     * @return true if file
     */
    public boolean isFile(FileInfo fileInfo) {
        return !EnumWithValue.EnumUtils.isSet(fileInfo.getFileAttributes(), FILE_ATTRIBUTE_DIRECTORY);
    }

    /**
     *
     * @param fileInfo
     * @return true if directory
     */
    public boolean isDirectory(FileInfo fileInfo) {
        return EnumWithValue.EnumUtils.isSet(fileInfo.getFileAttributes(), FILE_ATTRIBUTE_DIRECTORY);
    }

    /**
     *
     * @param path
     * @throws TransportException
     */
    public void delete(String path) throws TransportException {
        diskShare.rm(transformPath(path));
    }

    /**
     *
     * @param fileInfo
     * @throws TransportException
     */
    public void delete(FileInfo fileInfo) throws TransportException {
        delete(fileInfo.getFileName());
    }

    /**
     *
     * @param path
     * @return list of files
     * @throws SMBApiException
     * @throws TransportException
     */
    public List<FileInfo> list(String path) throws SMBApiException, TransportException {
        return diskShare.list(transformPath(path));
    }

    /**
     *
     * @param remotePath
     * @return InputStream
     * @throws IOException
     * @throws SMBApiException
     */
    public InputStream read(String remotePath) throws IOException, SMBApiException {
        return readBytes(transformPath(remotePath));
    }

    /**
     *
     * @param fileInfo
     * @return InputStream
     * @throws IOException
     * @throws SMBApiException
     */
    public InputStream read(FileInfo fileInfo) throws IOException, SMBApiException {
        return read(fileInfo.getFileName());
    }

    /**
     *
     * @param fileInfo
     * @param os
     * @throws IOException
     * @throws SMBApiException
     */
    public void read(FileInfo fileInfo, OutputStream os) throws IOException, SMBApiException {
        read(fileInfo.getFileName(), os);
    }

    /**
     *
     * @param remotePath
     * @param os
     * @throws IOException
     * @throws SMBApiException
     */
    public void read(String remotePath, OutputStream os) throws IOException, SMBApiException {
        com.hierynomus.smbj.share.File file = diskShare.openFile(transformPath(remotePath), EnumSet.of(GENERIC_READ), FILE_OPEN);
        file.read(os);
        file.close();
    }

    /**
     *
     * @param fileInfo
     * @param remotePath
     */
    public void remoteCopy(FileInfo fileInfo, String remotePath) {
        remoteCopy(fileInfo.getFileName(), remotePath);
    }

    /**
     *
     * @param remoteSourcePath
     * @param remoteDestinationPath
     */
    public void remoteCopy(String remoteSourcePath, String remoteDestinationPath) {
        try (InputStream stream = readBytes(transformPath(remoteSourcePath))) {
            write(transformPath(remoteDestinationPath), stream);
        } catch (IOException | SMBApiException ex) {
            log.error("Error while copy file " + remoteSourcePath, ex);
        }
    }

    /**
     *
     * @param remotePath
     * @param is
     * @throws IOException
     * @throws SMBApiException
     */
    public void write(String remotePath, InputStream is) throws IOException, SMBApiException {
        com.hierynomus.smbj.share.File file = diskShare.openFile(transformPath(remotePath), EnumSet.of(GENERIC_WRITE), FILE_OVERWRITE_IF);
        file.write(new InputStreamByteChunkProvider(is));
        file.close();
    }

    /**
     *
     * @param remotePath
     * @return OutputStream
     * @throws IOException
     * @throws SMBApiException
     */
    public OutputStream write(String remotePath) throws IOException, SMBApiException {
        com.hierynomus.smbj.share.File file = diskShare.openFile(transformPath(remotePath), EnumSet.of(GENERIC_WRITE), FILE_OVERWRITE_IF);
        return file.getOutputStream();
    }

    /**
     *
     * @param path - Remote file path
     * @return true if exist or false
     */
    public boolean exists(String path) {
        return diskShare.fileExists(transformPath(path));
    }

    /**
     *
     * @return true if exist or false. Smb url path exists
     */
    public boolean folderExists() {
        return diskShare.folderExists(transformPath(smbUrl.getPath()));
    }

    /**
     *
     * @param path
     * @return true if exist or false. Smb path exists
     */
    public boolean folderExists(String path) {
        return diskShare.folderExists(transformPath(path));
    }

    /**
     *
     * @param path
     * @throws TransportException
     */
    public void makeFolder(String path) throws TransportException {
        diskShare.mkdir(transformPath(path));
    }

    private InputStream readBytes(String remotePath) throws IOException, SMBApiException {
        com.hierynomus.smbj.share.File file = diskShare.openFile(transformPath(remotePath), EnumSet.of(GENERIC_READ), FILE_OPEN);
        ByteArrayInputStream byteStream = new ByteArrayInputStream(IOUtil.getBytes(file.getInputStream()));
        file.close();
        return byteStream;
    }

    private String transformPath(String path) {
        return path.replace('/', '\\');
    }
}
public class SmbUrl {

    private final URL smbUrl;

    public SmbUrl(String url) throws MalformedURLException {
        smbUrl = new URL(null, url, new SmbHandler());
    }

    public String getShare() {
        return Arrays
                .stream(smbUrl.getPath().split("/"))
                .filter(s -> s.length() > 0)
                .findFirst()
                .orElse("");
    }

    public String getPath() {
        String path = Arrays
                .stream(smbUrl.getPath().split("/"))
                .filter(s -> s.length() > 0)
                .skip(1)
                .collect(Collectors.joining("\\"));
        return !path.isEmpty() ? path + '\\' : path;
    }

    public int getPort() {
        return smbUrl.getPort();
    }

    public String getProtocol() {
        return smbUrl.getProtocol();
    }

    public String getHost() {
        return smbUrl.getHost();
    }

    @Override
    public String toString() {
        return smbUrl.toString();
    }

}
```
and simply call
```java
SMBClientEx client = new SMBClientEx("smb://host/share/directory1/directory2/directory3");
client.authenticate(new AuthenticationContext(user, password, domain);
```
How to call SMBClient in new version?
Thanks

**Comment by hierynomus** [comment]
In order to enable DFS lookups, you need to pass a Config into the constructor.

```
DefaultConfig cfg = new DefaultConfig();
cfg.setDFSEnabled(true);
SMBClient client = new SMBClient(cfg);
```

**Comment by gidravlic** [comment]
Yes, I understood it, looked in the source. And after? If i try connect with UNC path exception throw:
`java.net.UnknownHostException: \\mbtest\dfs-test 
at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)` and it is correct. 
At what stage is parsing the UNC path?

**Comment by hierynomus** [comment]
Can you post the entire stacktrace? It seems like it tries to use the `hostname/sharename` to resolve the IP address... Which is not correct.

**Comment by gidravlic** [comment]
In my test domain i don't use IPv6. Is it possible to prescribe a priority to IPv4?

`INFO: No error handler has been configured to handle exceptions during publication.
Publication error handlers can be added by IBusConfiguration.addPublicationErrorHandler()
Falling back to console logger.
13:41:42.191 [main] ERROR org.helper.smbdfstest.SMB - Erorr
java.net.UnknownHostException: \\mbtest.mbdomain.com\dfs-test
        at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
        at java.net.InetAddress$2.lookupAllHostAddr(Unknown Source)
        at java.net.InetAddress.getAddressesFromNameService(Unknown Source)
        at java.net.InetAddress.getAllByName0(Unknown Source)
        at java.net.InetAddress.getAllByName(Unknown Source)
        at java.net.InetAddress.getAllByName(Unknown Source)
        at java.net.InetAddress.getByName(Unknown Source)
        at com.hierynomus.protocol.commons.socket.SocketClient.connect(SocketClient.java:44)
        at com.hierynomus.smbj.SMBClient.getEstablishedOrConnect(SMBClient.java:78)
        at com.hierynomus.smbj.SMBClient.connect(SMBClient.java:58)
        at org.helper.smbdfstest.SMB.main(SMB.java:49)`

**Comment by hierynomus** [comment]
It seems you're calling `connect` with `\\mbtest.mbdomain.com\dfs-test` as the host name yourself.

You're looking at internal JRE code. You can tweak this by setting `-Djava.net.preferIPv6Addresses=false` on the command line. However you should be calling the `connect` method with `mbtest.mbdomain.com` without the `/dfs-test`, then it comprises a host name.

**Comment by gidravlic** [comment]
It seemed to me that it would be more logical to connect to the full path of DFS. As I understood the mechanism of connection to the host remained the same. Also what then has changed? I mean supporting DFS. Only that now we can connect to the DFS?

**Comment by hierynomus** [comment]
It should now be the case that the DFS referrals are transparent.

The sequence though is still:

```
DefaultConfig cfg = new DefaultConfig();
cfg.setDFSEnabled(true);
SMBClient client = new SMBClient(cfg);
Connection c = client.connect("mbtest.mbdomain.com");
Session session = c.authenticate(new AuthenticationContext(username, password, "dfs-test"));
session.connectShare(...);
```

**Comment by gidravlic** [comment]
> `Session session = c.authenticate(new AuthenticationContext(username, password, "dfs-test"));`
No, No. The third parameter is domain, not share name. One more question. You wrote that the use of dfs  referrals are transparent. What you mean? Please, show some example.

**Comment by hierynomus** [comment]
You're right, my bad. You need to pass `dfs-test` into `session.connectShare("dfs-test")`.

The DFS protocol in SMB2 ensures that you don't need to know where the data is stored, ie. you can just connect to the DFS server and request a file. The DFS referral protocol then does a lookup in the DFS namespace and connects to the right server that actually hosts the file.

With the current master version of SMBJ that is not possible.

**Comment by gidravlic** [comment]
Basically my test work.  If you have free time, please add some example as you see the use of the new functionality. Oh yes, i can show my test case.

**Comment by hierynomus** [comment]
@ckherrmann @sbellan Can one of you guys do a quick check also? The DFS support has been taking too long already ;)

**Comment by sbellan** [comment]
In my setup, the DFS namespace server ap01-qa(1.24) has a share space2, (\\ap01-qa\space2), which has folder which targets to a another node(1.25)

image

When I do,

    List<FileInfo> apac = share.list("node25\\subfolder");

It is able to send the dfs referral request and get a response,

request,
image

response,
image

But after that, it is sending the open request to dfs server(1.24)  itself and not 1.25 and fails with NOT_FOUND,

image

Its great that we have basic referral implementation, but still some integration.

**Comment by hierynomus** [comment]
@sbellan Can you get me the trace logs for the entire flow? Really need to setup a multi instance vm set here but that takes some time... :(

**Comment by sbellan** [comment]
@hierynomus Here are the debug logs and pcap file,



[dfs_archive.zip](https://github.com/hierynomus/smbj/files/1026652/dfs_archive.zip)

**Comment by hierynomus** [comment]
Thanks a lot!

Op 24 mei 2017 8:21 p.m. schreef "sbellan" <notifications@github.com>:

**Comment by sbellan** [comment]
Some other issues I noticed during testing,

By default we use JCEProvider and looks like MD4 is not supported by standard JCE and fails with,

2017-05-24 11:24:51.713 [main] DEBUG c.h.smbj.auth.NtlmAuthenticator - Received NTLM challenge from: CORP
Exception in thread "main" com.hierynomus.ntlm.NtlmException: com.hierynomus.security.SecurityException: java.security.NoSuchAlgorithmException: MD4 MessageDigest not available
	at com.hierynomus.ntlm.functions.NtlmFunctions.NTOWFv1(NtlmFunctions.java:96)
	at com.hierynomus.ntlm.functions.NtlmFunctions.NTOWFv2(NtlmFunctions.java:61)
	at com.hierynomus.smbj.auth.NtlmAuthenticator.authenticate(NtlmAuthenticator.java:90)
	at com.hierynomus.smbj.connection.Connection.authenticationRound(Connection.java:179)
	at com.hierynomus.smbj.connection.Connection.authenticate(Connection.java:149)
	at smbj.DfsTest.main(DfsTest.java:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: com.hierynomus.security.SecurityException: java.security.NoSuchAlgorithmException: MD4 MessageDigest not available
	at com.hierynomus.security.jce.JceMessageDigest.<init>(JceMessageDigest.java:38)
	at com.hierynomus.security.jce.JceSecurityProvider.getDigest(JceSecurityProvider.java:44)
	at com.hierynomus.ntlm.functions.NtlmFunctions.NTOWFv1(NtlmFunctions.java:92)
	... 10 more
Caused by: java.security.NoSuchAlgorithmException: MD4 MessageDigest not available
	at sun.security.jca.GetInstance.getInstance(GetInstance.java:159)
	at java.security.Security.getImpl(Security.java:695)
	at java.security.MessageDigest.getInstance(MessageDigest.java:167)
	at com.hierynomus.security.jce.JceMessageDigest.<init>(JceMessageDigest.java:35)
	... 12 more

When I used the BouncyCastle Provider there was a typo,
`diff --git a/src/main/java/com/hierynomus/security/bc/BCMac.java b/src/main/java/com/hierynomus/security/bc/BCMac.java
index 1c706ea..dfdd87d 100644
--- a/src/main/java/com/hierynomus/security/bc/BCMac.java
+++ b/src/main/java/com/hierynomus/security/bc/BCMac.java
@@ -28,7 +28,7 @@ import java.util.Map;
 public class BCMac implements Mac {
     private static Map<String, Factory<org.bouncycastle.crypto.Mac>> lookup = new HashMap<>();
     static {
-        lookup.put("HMACSHA265", new Factory<org.bouncycastle.crypto.Mac>() {
+        lookup.put("HMACSHA256", new Factory<org.bouncycastle.crypto.Mac>() {
             @Override
             public org.bouncycastle.crypto.Mac create() {
                 return new HMac(new SHA256Digest());`

**Comment by hierynomus** [comment]
@sbellan Fixed those two issues. You can now use JCE also. Java has an MD4 implementation, it just is not available by default in the JCE. 

Still working on the DFS bug. I figured out what I missed... Just need to re-think part of the architecture ;)

**Comment by Disablez** [comment]
Currently this branch causes an exception when connecting due to the "data" array in Buffer.java being null

Exception in thread "main" java.lang.NullPointerException
	at com.hierynomus.protocol.commons.buffer.Buffer.<init>(Buffer.java:106)
	at com.hierynomus.protocol.commons.buffer.Buffer.<init>(Buffer.java:95)
	at com.hierynomus.protocol.commons.buffer.Buffer$PlainBuffer.<init>(Buffer.java:49)
	at com.hierynomus.smbj.auth.NtlmAuthenticator.authenticate(NtlmAuthenticator.java:81)
	at com.hierynomus.smbj.connection.Connection.authenticate(Connection.java:163)
	at com.jp.SMBAuthenticatorSmbj.connect(Main.java:77)
	at com.jp.Main.main(Main.java:38)

This doesn't happen on main branch, so I guess I need to try and make a collage of both branches for the time being.

**Comment by hierynomus** [comment]
This branch is not yet finished, and the DFS support does not work (ie.
resolution works, but then it does not use the new connection).

2017-06-26 11:26 GMT+02:00 Disablez <notifications@github.com>:

File: src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java [hunk 1]
```diff
@@ -61,82 +63,84 @@ public NtlmAuthenticator create() {
...
+                    return negTokenTarg(resp, negTokenTarg.getResponseToken());
+                } else {
+                    NtlmAuthenticate resp = new NtlmAuthenticate(new byte[0], ntlmv2Response,
+                        context.getUsername(), context.getDomain(), null, sessionkey, EnumWithValue.EnumUtils.toLong(negotiateFlags),
+                        false
+                    );
+                    return negTokenTarg(resp, negTokenTarg.getResponseToken());
+                }
             }
+        } catch (SpnegoException spne) {
```
Comment (srik545) [inline src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java]: Hi Guys, Thanks for the hardwork. One question.
Just trying to understand, why in NTLMAuthentication, SpnegoException is thrown?

**Comment by srik545** [comment]
Hi @hierynomus,
I have some time to test DFS share access with live servers. But I am getting below exception. 
Any idea when this branch can be used for testing?

Exception in thread "main" java.lang.NullPointerException
	at com.hierynomus.protocol.commons.buffer.Buffer.<init>(Buffer.java:106)
	at com.hierynomus.protocol.commons.buffer.Buffer.<init>(Buffer.java:95)
	at com.hierynomus.protocol.commons.buffer.Buffer$PlainBuffer.<init>(Buffer.java:49)
	at com.hierynomus.smbj.auth.NtlmAuthenticator.authenticate(NtlmAuthenticator.java:81)
	at com.hierynomus.smbj.connection.Connection.authenticate(Connection.java:169)

-Thanks.

**Comment by hierynomus** [comment]
@Disablez @srik545 The bug in the authentication was due to a merge gone wrong. That is now fixed.

**Comment by hierynomus** [comment]
@pepijnve Do you have a DFS setup somewhere you can try this PR on? I tested it on my setup, which now seems to work.

**Comment by pepijnve** [comment]
I have a NetApp with widelinks configured. That uses DFS referrals. I'll see if I can test against that.

**Comment by pepijnve** [comment]
I did a really quick check by updating the DfsIntegrationSpec for the test server and stepped through the DFS resolution process with the debugger. Worked perfectly. :+1:

**Comment by hierynomus** [comment]
Thanks @pepijnve!

**Pull request merged**